packages/Kokkos/src/DTK_UserApplication.hpp

```diff
@@ -0,0 +1,157 @@
+  names of its contributors may be used to endorse or promote products
+  derived from this software without specific prior written permission.
+
+  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+  HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+//---------------------------------------------------------------------------//
+/*!
+ * \brief DTK_UserApplication.hpp
+ * \author Stuart R. Slattery
+ * \brief Interface to user applications.
+ */
+//---------------------------------------------------------------------------//
+
+#ifndef DTK_USERAPPLICATION_HPP
+#define DTK_USERAPPLICATION_HPP
+
+#include "DTK_Boundary.hpp"
+#include "DTK_BoundingVolumeList.hpp"
+#include "DTK_CellList.hpp"
+#include "DTK_ConfigDefs.hpp"
+#include "DTK_DofIdMap.hpp"
+#include "DTK_EvaluationSet.hpp"
+#include "DTK_Field.hpp"
+#include "DTK_NodeList.hpp"
+#include "DTK_ParallelTraits.hpp"
+#include "DTK_PolyhedronList.hpp"
+#include "DTK_UserFunctionRegistry.hpp"
+#include "DTK_View.hpp"
+
+#include <Kokkos_Core.hpp>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace DataTransferKit
+{
+//---------------------------------------------------------------------------//
+/*!
+ * \class UserApplication
+ *
+ * \brief High-level interface to user applications.
+ *
+ * \tparam Scalar Scalar type of the fields this user application will
+ * provide.
+ *
+ * \tparam ParallelModel The DTK parallel model indicating the on-node
+ * parallelism of the user application. Indicates where data will be
+ * allocated.
+ *
+ * The user application provides a high-level interface to compose DTK input
+ * data structures and push and pull field data to and from the application
+ * through sequences of user function calls.
+ */
+//---------------------------------------------------------------------------//
+template <class Scalar, class ParallelModel>
+class UserApplication
+{
+  public:
+    //@{
+    //! Type aliases.
+    using ExecutionSpace =
+        typename ParallelTraits<ParallelModel>::ExecutionSpace;
+    //@}
+
+    // Constructor.
+    UserApplication(
+        const std::shared_ptr<UserFunctionRegistry<Scalar>> &user_functions );
+
+    // Get a node list from the application.
+    NodeList<Kokkos::LayoutLeft, ExecutionSpace> nodeList();
```

Comment:
Right - we set Kokkos::LayoutLeft to ensure a strict ordering in our 1D user interfaces. We can discuss names